File: 0740. Delete and Earn/Solution0740.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

class Solution0740 {
public:
    int deleteAndEarn(vector<int>& nums) {
        int maxNum = 0;
        for (int i = 0; i < nums.size(); i++) {
            maxNum = max(maxNum, nums[i]);
        }
        vector<int> sums(maxNum+1, 0);        
        vector<int> dp(maxNum+1, 0);
        for (int i = 0; i < nums.size(); i++) {
            sums[nums[i]] += nums[i];
        }
        dp[1] += sums[1];
        for (int i = 2; i < maxNum+1; i++) {
            dp[i] = max(dp[i-2] + sums[i], dp[i-1]);
        }
        return dp[maxNum];
    }
};

int main() {
    Solution0740 sol;
    vector<int> nums{3,4,2};
    int res = sol.deleteAndEarn(nums);
    cout << res << endl;
}

File: 0740. Delete and Earn/Solution0740.py
from typing import List


class Solution0740:
    def deleteAndEarn(self, nums: List[int]) -> int:
        max_num = max(nums)
        sums = [0]* (max_num + 1)
        for num in nums:
            sums[num] += num 
        dp = [0]* (max_num + 1)
        
        dp[1] += sums[1]
        
        for i in range(2, max_num + 1):
            dp[i] = max(dp[i-2] + sums[i], dp[i-1])
        
        return dp[max_num]